File: solutions/1056-capacity-to-ship-packages-within-d-days/solution.java
class Solution {
    public static int max(int[] arr){
        int max = arr[0];
        for (int i = 0; i < arr.length; i++) if(max<arr[i]) max = arr[i];
        return max;
    }
    public static int sum(int[] arr){
        int sum = 0;
        for (int element : arr) sum+=element;
        return sum;
    }
    public static long timeCalculation(int[] arr, int cap){
        long sum = 0;
        int day = 0;
        int i =0;
        while (i< arr.length){
            sum = sum+arr[i];
            if(sum==cap) {
                day++;sum=0;
            }
            else if(sum>cap) {
                sum = 0;
                i--;day++;
            }
            else if (i == arr.length - 1 && sum!=0 && sum < cap) day++;
            i++;
        }
        return day;
    }
    public static int shipWithinDays(int[] weights, int days) {
        int low = max(weights);
        int high = sum(weights)*2/days;
        while(low<=high){
            int mid = low+(high-low)/2;
            long time = timeCalculation(weights, mid);
            if(time> days) low = mid+1;
            else high = mid-1;
        }
        return low;
    }
}

File: solutions/1056-capacity-to-ship-packages-within-d-days/solution.cpp
class Solution {
public:
    int daycal(vector<int>& weights, int capacity){
        int sum = 0;
        int cntdays = 0;
        for(int ele: weights){
            sum+=ele;
            if(sum>capacity){
                sum = ele;
                cntdays++;
            }
        }
        if(sum==0) return cntdays;
        else return cntdays+1;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int maxi = 0;
        int sum = 0;
        for(int ele: weights) maxi = max(maxi, ele);
        for(int ele: weights) sum+=ele;
        int l = maxi;
        int r = sum;
        int ans = sum;
        while(l<r){
            int midcap = l+(r-l)/2;
            int day = daycal(weights, midcap);
            // cout<<midcap<<" "<<l<<" "<<r<<" "<<day<<endl;
            if(day>days) {
                l = midcap+1;
            }
            else {
                ans = min(ans, midcap);
                r = midcap;
            }
        }
        return ans;
    }
};
